### PR TITLE
Proper container configurator API

### PR DESCRIPTION
**Pull request by Wirone**
As discussed in [#7035](https://github.com/rectorphp/rector/discussions/7035) here's implementation of `RectorServiceConfigurator` and rector for migrating from Symfony's one.

## Rector rule

I've tested it on Rector's codebase and it works, but it's unnecessary to refactor internal code since `symfony/dependency-injection` is patched and locally has proper return type. Also I did not add `UseRectorContainerConfiguratorRector` to any rule set because I think it should be opt-in for end users (explicit use in the config). But if you would like to add it, just let me know where it should be.

### CODE REVIEW

**Comment by Wirone** [comment]
@TomasVotruba any news on this? Is there any blocker that stop you from merging this?

File: ecs.php [hunk 1]
```diff
@@ -40,6 +40,9 @@
     // make ECS 16x faster
     $parameters->set(Option::PARALLEL, true);
 
+    // Reusable ECS cache for Docker runtime
+    $parameters->set(Option::CACHE_DIRECTORY, 'tmp/ecs');
+
```
Comment (TomasVotruba) [inline ecs.php]: Could you restore the defaults? This now conflicts with non-docker usage

File: docker-compose.yml [hunk 2]
```diff
@@ -4,6 +4,9 @@ services:
   php:
     build: .
     environment:
-        - COMPOSER_ROOT_VERSION=dev-main
+        COMPOSER_ROOT_VERSION: dev-main
+        PHP_IDE_CONFIG: "serverName=${PHP_IDE_CONFIG:-PHP_Docker}"
+        XDEBUG_MODE: ${XDEBUG_MODE:-off}
+        XDEBUG_CONFIG: 'idekey=${XDEBUG_IDEKEY:-PHPSTORM} client_host=${XDEBUG_CLIENT_HOST:-host.docker.internal}'
     volumes:
```
Comment (TomasVotruba) [inline docker-compose.yml]: Could you extract Docker changes to standalone PR? It's different change unrelated to custom Container extension.

File: rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+?>
+-----
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Closure\UseRectorContainerConfiguratorRector\Fixture;
+
+return static function (\Rector\Core\DependencyInjection\Loader\Configurator\RectorContainerConfigurator $containerConfigurator): void {
+};
+
+?>
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc:17]: Could you add a test fixture to skip the other non-rector configs?
We don't want to override Symfony `config/services.php` to Rector ones :) also ECS should be skipped.

File: src/DependencyInjection/Loader/Configurator/RectorContainerConfigurator.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+
+/**
+ * @api
+ * Same as Symfony container configurator, with patched return type for "set()" method for easier DX.
+ * It is an alias for internal class that is prefixed during build, so it's basically for keeping stable public API.
+ */
+final class RectorContainerConfigurator extends ContainerConfigurator
+{
+}
```
Comment (TomasVotruba) [inline src/DependencyInjection/Loader/Configurator/RectorContainerConfigurator.php:16]: I could not find the test case, that would actually test this type in `rector.php` results in successfuly loaded Rector container.
Could you add one?

**Comment by TomasVotruba** [comment]
@Wirone I'm a bit ill :face_with_thermometer:  this week, so I only got to your PR.
I've dropped few comments :+1:

File: ecs.php (hunk 1 above)
Comment (Wirone) [reply ecs.php]: @TomasVotruba with Docker cache only makes sense with explicitly specified path, because it needs to be mounted as volume. Otherwise every run will start from scratch... And it's not safe to specify path outside of the project (absolute path) because you can't be sure which path is always available.

File: docker-compose.yml (hunk 2 above)
Comment (Wirone) [reply docker-compose.yml]: OK, I'll extract it to another PR.

File: rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (Wirone) [reply rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc:17]: How it should be checked? IMHO there's no need for such fixtures since this rule is optional (I did not add it to any rule set). People who want opt-in and use it should just configure paths to ignore for this rule.

File: docker-compose.yml (hunk 2 above)
Comment (Wirone) [reply docker-compose.yml]: It's in #1928

File: ecs.php (hunk 1 above)
Comment (Wirone) [reply ecs.php]: Moved to #1928

File: rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (TomasVotruba) [reply rules-tests/TypeDeclaration/Rector/Closure/UseRectorContainerConfiguratorRector/Fixture/fixture.php.inc:17]: Probably e2e test, that single registered configured rule did some job.

**Comment by TomasVotruba** [comment]
I'll look into this one today :+1:

**Comment by TomasVotruba** [comment]
@Wirone Thank you for your work on this :+1: 

Continue in https://github.com/rectorphp/rector-src/pull/2019
